indra/newview/llagent.cpp

```diff
@@ -1459,69 +1459,64 @@ LLVector3 LLAgent::getReferenceUpVector()
-
-    // don't let user pitch if pointed almost all the way down or up
-    LLVector3 skyward = getReferenceUpVector();
+    if (gAgentCamera.getCameraMode() == CAMERA_MODE_THIRD_PERSON ||
+        gAgentCamera.getCameraMode() == CAMERA_MODE_MOUSELOOK)
+    {
+        // Backup the current orientation
+        LLCoordFrame saved_frame_agent(mFrameAgent);
 
-    // A dot B = mag(A) * mag(B) * cos(angle between A and B)
-    // so... cos(angle between A and B) = A dot B / mag(A) / mag(B)
-    //                                  = A dot B for unit vectors
-    F32 agent_camera_angle_from_skyward = acos(newCoordFrame.getAtAxis() * skyward) * RAD_TO_DEG;
+        // Optimistic rotation up/down (vertical angle can reach and exceed 0 or 180)
+        mFrameAgent.pitch(angle);
 
-    F32 min_angle = 1;
-    F32 max_angle = 179;
-    bool check_viewer_camera = false;
+        // Cosine of the angle between current agent At and Up directions
+        F32 agent_at_to_up_now_cos = mFrameAgent.mXAxis * gAgentCamera.getCameraUpVector();
+        bool pitch_away_from_horizont = (angle < 0) ^ (agent_at_to_up_now_cos < 0);
+        // We always allow to pitch in direction to horizont (from zenith or from nadir)
+        if (!pitch_away_from_horizont)
+            return;
 
-    if (gAgentCamera.getCameraMode() == CAMERA_MODE_THIRD_PERSON)
-    {
-        // These values of min_angle and max_angle are obtained purely empirically
-        if (gAgentCamera.getCameraPreset() == CAMERA_PRESET_REAR_VIEW)
+        // Current angle between agent At and Up directions
+        F32 agent_at_to_up_now = acos(agent_at_to_up_now_cos);
+        // Requested angle between agent At and Up directions
+        F32 agent_at_to_up_new = agent_at_to_up_now + angle;
+        F32 agent_at_to_up_new_sin = sin(agent_at_to_up_new);
+        // Overpitched? Then rollback
+        if (agent_at_to_up_new_sin < 1e-4)
         {
-            min_angle = 10;
-            check_viewer_camera = true;
+            mFrameAgent = saved_frame_agent;
+            return;
         }
-        else if (gAgentCamera.getCameraPreset() == CAMERA_PRESET_GROUP_VIEW)
+
+        if (gAgentCamera.getCameraMode() == CAMERA_MODE_THIRD_PERSON)
         {
-            min_angle = 10;
-            max_angle = 170;
-            check_viewer_camera = true;
+            // Camera sight relative to agent frame (focus - offset)
+            LLVector3 camera_offset(gAgentCamera.getCameraOffsetInitial());
+            LLVector3 camera_focus(gAgentCamera.getFocusOffsetInitial());
+            LLVector3 camera_sight(camera_focus - camera_offset);
+            // 2D projection of the camera sight to the XZ plane
+            LLVector2 camera_sight_2d_vert(1, camera_sight[VZ]);
+            camera_sight_2d_vert.normalize();
+            // Cosine of the 2D angle between initial camera At and X axis (in the XZ plane)
+            F32 camera_sight_to_at_2d_vert_cos = camera_sight_2d_vert * LLVector2(LLVector3::x_axis);
```

Comment:
I'm not sure XZ plane is the right chose. 'plane' is relative. Ex: If you are flying a vehicle vertically or upside down...

Reply:
I've meant an XZ plane in agent coordinate system, not in world
In the THIRD_PERSON mode the viewer camera and the agent camera have the constant vertical angle between each other
For example, if it is 10 degrees, then when agent looks up 80 degrees over horizont, the viewer camera looks at zenith
This code just calculates this vertical angle between the viewer camera and the agent camera